### PR TITLE
[Merged by Bors] - Detailed validator monitoring

### LINKED ISSUE #2064: Add validator balance Prometheus metric
## Description

Seems to be solid demand for this in Discord, some users have even made their own workarounds.

## Version

1.0.3

## Steps to resolve

Seems like the easiest way would be to query the standard API for balances on startup and then once per epoch for all enabled validators. And only do this if the VC has the `--metrics` flag.

It'd be adding some non-functional overhead by making completely new beacon node requests but I think if people are looking for validator metrics, this is one of the main one's they're after.

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

- Resolves #2064

## Proposed Changes

Adds a `ValidatorMonitor` struct which provides additional logging and Grafana metrics for specific validators.

Use `lighthouse bn --validator-monitor` to automatically enable monitoring for any validator that hits the [subnet subscription](https://ethereum.github.io/eth2.0-APIs/#/Validator/prepareBeaconCommitteeSubnet) HTTP API endpoint.

Also, use `lighthouse bn --validator-monitor-pubkeys` to supply a list of validators which will always be monitored.

See the new docs included in this PR for more info.

## TODO

- [x] Track validator balance, `slashed` status, etc.
- [x] ~~Register slashings in current epoch, not offense epoch~~
- [ ] Publish Grafana dashboard, update TODO link in docs
- [x] ~~#2130 is merged into this branch, resolve that~~

### CODE REVIEW

**Review by realbigsean — CHANGES_REQUESTED** [review]
This is awesome! Can't wait till it's in :rocket: 

I left a bunch of comments, most of them are just typos. The only sort of major thing I would be concerned about is opening a read lock on `ValidatorMonitor` whenever we register something, and subsequently opening a write lock on the `SummaryMap` (which is held in the `ValidatorMonitor` struct) during this registration. I'm not exactly sure if/how it could deadlock but wanted to bring it up

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 1]
```diff
@@ -0,0 +1,986 @@
...
+    fn get_aggregated_attestation_delay_ms<S: SlotClock>(
+        seen_timestamp: Duration,
+        data: &AttestationData,
+        slot_clock: &S,
+    ) -> Duration {
+        slot_clock
+            .start_of(data.slot)
+            .and_then(|slot_start| seen_timestamp.checked_sub(slot_start))
+            .and_then(|gross_delay| {
+                let production_delay = slot_clock.slot_duration() / 2;
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/validator_monitor.rs:545]: Great catch!

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 2]
```diff
@@ -0,0 +1,986 @@
...
+                if let Some(validator) = state.validators.get(i) {
+                    metrics::set_int_gauge(
+                        &metrics::VALIDATOR_MONITOR_EFFECTIVE_BALANCE_GWEI,
+                        &[id],
+                        u64_to_i64(validator.effective_balance),
+                    );
+                    metrics::set_int_gauge(
+                        &metrics::VALIDATOR_MONITOR_SLASHED,
+                        &[id],
+                        if validator.slashed { 0 } else { 1 },
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/validator_monitor.rs]: Good catch!

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 3]
```diff
@@ -0,0 +1,986 @@
...
+    ///
+    /// ## Notes
+    ///
+    /// - If `epoch` doesn't exist in `self.summaries`, it is created.
+    /// - `self.summaries` may be pruned after `func` is run.
+    fn with_epoch_summary<F>(&self, epoch: Epoch, func: F)
+    where
+        F: Fn(&mut EpochSummary),
+    {
+        let mut summaries = self.summaries.write();
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/validator_monitor.rs:163]: Thanks for pointing this out. I believe it's safe though.

The order of lock acquisition is always `lock(external) -> lock(internal) -> unlock(internal) -> unlock(external)`. We provide no method to expose the `internal` lock, so this always holds.

I've added a comment to warn the user of a potential internal deadlock with `with_epoch_summary`. though!

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 4]
```diff
@@ -0,0 +1,986 @@
...
+            validators: <_>::default(),
+            indices: <_>::default(),
+            auto_register,
+            log,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Add some validators to `self` for additional monitoring.
+    pub fn add_validators_from_file<P: AsRef<Path>>(&mut self, path: P) -> Result<(), Error> {
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/validator_monitor.rs]: Removed!

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 5]
```diff
@@ -0,0 +1,986 @@
...
+    pub fn register_api_attester_slashing(
+        &self,
+        seen_timestamp: Duration,
+        slashing: &AttesterSlashing<T>,
+    ) {
+        self.register_attester_slashing("api", seen_timestamp, slashing)
+    }
+
+    /// Register an attester slashing included in a *valid* `BeaconBlock`.
+    pub fn register_block_attester_slashing(
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/validator_monitor.rs]: I've fixed this here: https://github.com/sigp/lighthouse/pull/2151/commits/82cd92129671a4765de8fdff77317beaab573bf7

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 6]
```diff
@@ -0,0 +1,986 @@
...
+    pub fn register_api_proposer_slashing(
+        &self,
+        seen_timestamp: Duration,
+        slashing: &ProposerSlashing,
+    ) {
+        self.register_proposer_slashing("api", seen_timestamp, slashing)
+    }
+
+    /// Register a proposer slashing included in a *valid* `BeaconBlock`.
+    pub fn register_block_proposer_slashing(
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/validator_monitor.rs]: I've fixed this here: https://github.com/sigp/lighthouse/pull/2151/commits/82cd92129671a4765de8fdff77317beaab573bf7

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 7]
```diff
@@ -0,0 +1,986 @@
...
+
+    /// Add some validators to `self` for additional monitoring.
+    pub fn add_validator_pubkeys(&mut self, pubkeys: Vec<PublicKeyBytes>) {
+        for pubkey in pubkeys {
+            self.add_validator_pubkey(pubkey)
+        }
+    }
+
+    /// Add some validators to `self` for additional monitoring.
+    pub fn add_validator_pubkey(&mut self, pubkey: PublicKeyBytes) {
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/validator_monitor.rs]: In https://github.com/sigp/lighthouse/pull/2151/commits/bfcb6fe6981d4026ca00b758f34d83ef623f9af5 I've made it so `new` takes a list of validators and the `BeaconChainBuilder` adds the canonical head state to the validator monitor when it's being built.

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 8]
```diff
@@ -0,0 +1,986 @@
...
+        seen_timestamp: Duration,
+        data: &AttestationData,
+        slot_clock: &S,
+    ) -> Duration {
+        slot_clock
+            .start_of(data.slot)
+            .and_then(|slot_start| seen_timestamp.checked_sub(slot_start))
+            .and_then(|gross_delay| {
+                let production_delay = slot_clock.slot_duration() / 3;
+                gross_delay.checked_sub(production_delay)
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/validator_monitor.rs:419]: I agree, I think I might leave it for later works though. I'm not exactly sure how Prom/Graf will handle negative a duration and I'm keen to get this out the door.

**Comment by paulhauner** [comment]
All comments addressed! Thanks for the review, super helpful.

I added a couple of extra metrics in 16d1f089f, so I'll have to update the dashboard and replace the PR you made (sorry!).

**Review by realbigsean — APPROVED** [review]
Nice! Everything looks good

**Pull request closed without merging**